src/tools/clippy/clippy_lints/src/transmute.rs

```diff
@@ -648,3 +693,57 @@ fn is_layout_incompatible<'tcx>(cx: &LateContext<'tcx>, from: Ty<'tcx>, to: Ty<'
         false
     }
 }
+
+/// Check if the the type conversion can be expressed as a pointer cast, instead of a transmute.
+fn can_be_expressed_as_pointer_cast<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>, from_ty: Ty<'tcx>, to_ty: Ty<'tcx>) -> bool {
+    use CastKind::*;
+    matches!(
+        check_cast(cx, e, from_ty, to_ty),
+        Some(
+            PtrPtrCast
+            | PtrAddrCast
+            | AddrPtrCast
+            | ArrayPtrCast
+            | FnPtrPtrCast
+            | FnPtrAddrCast
+        )
+    )
+}
+
+/// If a cast from from_ty to to_ty is valid, returns an Ok containing the kind of the cast.
+fn check_cast<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>, from_ty: Ty<'tcx>, to_ty: Ty<'tcx>) -> Option<CastKind> {
+    let hir_id = e.hir_id;
+    let local_def_id = hir_id.owner;
+
+    Inherited::build(cx.tcx, local_def_id).enter(|inherited| {
+        let fn_ctxt = FnCtxt::new(
+            &inherited,
+            // TODO should we try to get the correct ParamEnv?
+            ty::ParamEnv::empty(),
+            hir_id
+        );
+
+        // If we already have errors, we can't be sure we can pointer cast.
```

Comment:
can this happen? Just the creation shouldn't set this flag, right? Maybe just `assert` here so we know if/when it can happen

Reply:
I have not observed this happening. Asserting here sounds good to me.